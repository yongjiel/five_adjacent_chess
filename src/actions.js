export const INCREMENT = "INCREMENT";
export const DECREMENT = "DECREMENT";
export const RESET = "RESET";
export const CLICK_GRID = 'CLICK_GRID';



export function increment() {
  return { type: INCREMENT };
}


export const decrement = () => ({ type: DECREMENT });

export const reset = () => ({ type: RESET });

export const click_grid = () => ({ type: CLICK_GRID });