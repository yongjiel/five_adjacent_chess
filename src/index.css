body {
  font: 14px "Century Gothic", Futura, sans-serif;
  margin: 20px;
}

ol,
ul {
  padding-left: 30px;
}

.board-row:after {
  clear: both;
  content: "";
  display: table;
}

.status {
  margin-bottom: 10px;
}

.square {
  background: #fff;
  border: 1px solid #999;
  float: left;
  font-size: 24px;
  font-weight: bold;
  line-height: 34px;
  height: 34px;
  margin-right: -1px;
  margin-top: -1px;
  padding: 0;
  text-align: center;
  width: 34px;
}

.square.blue {
  background: blue;
}

.square:focus {
  outline: none;
}

.kbd-navigation .square:focus {
  background: #ddd;
}

.game {
  display: flex;
  flex-direction: row;
}

.game-info {
  margin-left: 20px;
  margin-right: 20px;
  max-width: 20%;
  max-height: 20%;
}

.game-input {
  /*background: gold;*/
}

.game-board {
  float: left;
  max-width: 80%;
  max-height: 80%;
}

.header ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
  overflow: hidden;
  background-color: #333333;
}

.header li {
  float: left;
}

.header li a {
  display: block;
  color: rgb(245, 238, 238);
  text-align: center;
  padding: 16px;
  text-decoration: none;
}

.header li a:hover {
  background-color: #111111;
}

html,
body,
#root,
.app,
.room {
  height: 100%;
}
html,
body {
  margin: 0;
  padding: 0;
  font-family: Georgia, "Times New Roman", Times, serif;
}
.room {
  text-align: center;
  padding-top: 30px;
  font-size: 30px;
}
.room.lit {
  background: #fff;
  color: #222;
}
.room.dark {
  background: #222;
  color: white;
}
