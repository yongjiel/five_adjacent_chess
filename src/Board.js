import React from "react";
import Square from "./Square.js";
import { click_grid } from "./actions";
import { connect } from "react-redux";

class Board extends React.Component {
  constructor(props) {
    super(props);
    this.click_grid = this.click_grid.bind(this);
  }

  click_grid(i) {
    this.props.click_grid(i);
  }

  renderSquare(i, is_bule_color) {
    return (
      <Square
        stepNumber={this.props.stepNumber}
        is_bule_color={is_bule_color}
        value={this.props.history[this.props.stepNumber].squares[i]}
        onClick={() => this.click_grid(i)}
        key={"Square#" + i}
      />
    );
  }

  render() {
    // this part is crazy good for loop and close div tag
    const match = this.props.match;
    const row_c = this.props.rows;
    var rows = [];
    for (var i = 0; i < row_c; i++) {
      rows.push(<div className="board-row" key={"row#" + i} />);
      for (var j = 0; j < row_c; j++) {
        let n = i * row_c + j;
        rows.push(this.renderSquare(n, match.indexOf(n) > -1));
      }
    }

    return <div>{rows}</div>;
  }
}

function mapStateToProps(store) {
  return {
    history: store.reducer.history, // initiate store.reducer of squares of the board and keep all the steps' states of all squares.
    stepNumber: store.reducer.stepNumber, // current step number.
    rows: store.reducer.rows, // also is the number of columns.
    match: store.reducer.match, // hold the square indices(linear in array) when win the game.
  };
}

// in this object, keys become prop names,
// and values should be action creator functions.
// They get bound to `dispatch`.
const mapDispatchToProps = {
  click_grid,
};

export default connect(mapStateToProps, mapDispatchToProps)(Board);
