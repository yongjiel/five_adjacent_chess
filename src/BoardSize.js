import React from "react";
import { input_size } from "./actions.js";
import { connect } from "react-redux";
import Rows from "./square_rows.js";

class BoardSize extends React.Component {
  constructor(props) {
    super(props);
    this.input_size = this.input_size.bind(this);
  }

  input_size(size) {
    this.props.input_size(size);
  }

  componentDidMount() {
    if (!!this.props.rows_ && !this.props.fixed) {
      this.input_size(parseInt(this.props.rows_));
    }
  }

  render() {
    return (
      <div className="game-input" key={"rows"}>
        <Rows
          rows={this.props.rows}
          onChange={(size) => this.input_size(size)}
          fixed={this.props.fixed}
        />
      </div>
    );
  }
}

function mapStateToProps(store) {
  return {
    rows: store.reducer.rows, // also is the number of columns.
    fixed: store.reducer.fixed,
  };
}

// in this object, keys become prop names,
// and values should be action creator functions.
// They get bound to `dispatch`.
const mapDispatchToProps = {
  input_size,
};

export default connect(mapStateToProps, mapDispatchToProps)(BoardSize);
