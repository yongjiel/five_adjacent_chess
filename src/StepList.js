import React from "react";
import { connect } from "react-redux";
import Step from "./Step.js";
import { click_step, click_toggle } from "./actions";

class StepList extends React.Component {
  constructor(props) {
    super(props);
    this.click_step = this.click_step.bind(this);
    this.click_toggle = this.click_toggle.bind(this);
  }

  click_step(i) {
    this.props.click_step(i);
  }

  click_toggle() {
    this.props.click_toggle();
  }

  renderStep(i, history_step, stepNumber, current_location, order) {
    return (
      <Step
        history_step={history_step}
        order={order}
        current_location={current_location}
        current_step={stepNumber}
        i={i}
        onClick={() => this.click_step(i)}
        key={"Step#" + i}
      />
    );
  }

  render() {
    // this part is crazy good for loop and close div tag
    const history = this.props.history;
    const row_c = history.length;
    var steps = [];
    var locations = this.props.locations;
    for (var j = 0; j < row_c; j++) {
      steps.push(
        <li key={j} className="step_list">
          {this.renderStep(
            j,
            history[j],
            this.props.stepNumber,
            locations[j],
            this.props.order
          )}
        </li>
      );
    }
    var reverse = false;
    if (this.props.order === "Descend") {
      steps = steps.reverse();
      reverse = true;
    }
    return (
      <div>
        <ul>
          <button onClick={() => this.click_toggle()}>
            <b>{this.props.order}</b>
          </button>
        </ul>
        <ol reversed={reverse}>{steps}</ol>
      </div>
    );
  }
}

function mapStateToProps(store) {
  return {
    history: store.reducer.history, // initiate store.reducer of squares of the board and keep all the steps' states of all squares.
    stepNumber: store.reducer.stepNumber, // current step number.
    locations: store.reducer.locations,
    order: store.reducer.order,
  };
}

// in this object, keys become prop names,
// and values should be action creator functions.
// They get bound to `dispatch`.
const mapDispatchToProps = {
  click_step,
  click_toggle,
};

export default connect(mapStateToProps, mapDispatchToProps)(StepList);
