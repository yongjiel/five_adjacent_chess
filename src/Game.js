import React from "react";
import Board from "./Board.js";
import BoardSize from "./BoardSize.js";
import StepList from "./StepList.js";
import { connect } from "react-redux";
import { reset } from "./actions";

export class Game extends React.Component {
  constructor(props) {
    super(props);
    this.restart = this.restart.bind(this);
  }

  restart() {
    this.props.reset();
  }

  render() {
    // When using bind in updateInputValue function, this can be used inside updateInputValue()

    // onClick={(i) => this.click_grid(i)} is amazing. Here it knows the index
    // of the square array in Board.js. Board.js has defined the onClick for Game.js's onClick
    // function with arg i.
    return (
      <div key={"Game1"}>
        <BoardSize rows_={this.props.rows_} /> <br />
        <div className="game">
          <Board />

          <div className="game-info">
            <div className="game-reset">
              <ul>
                <button className="reset" onClick={() => this.restart()}>
                  Restart the Game
                </button>
              </ul>
            </div>
            <div
              className="winner"
              style={{
                color: this.props.win_color,
                fontWeight: this.props.bold2,
              }}
            >
              {this.props.status}
            </div>
            <StepList />
          </div>
        </div>
      </div>
    );
  }
}

// Add this function:
// Crucial point. store is a multi reducers collection.
// Need to point out which reducer to use.
export function mapStateToProps(store) {
  return {
    status: store.reducer.status,
    win_color: store.reducer.win_color,
    bold2: store.reducer.bold2,
  };
}

// in this object, keys become prop names,
// and values should be action creator functions.
// They get bound to `dispatch`.
// const mapDispatchToProps = {
// reset,
// };

export const mapDispatchToProps = (dispatch) => ({
  reset: () => dispatch(reset()),
});

// Must export!
//export default Game;
const ConnectedGame = connect(mapStateToProps, mapDispatchToProps)(Game);
export default ConnectedGame;
