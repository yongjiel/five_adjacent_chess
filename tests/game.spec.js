import React from "react";
//import { expect } from "chai";  // use jest expect
import Adapter from "enzyme-adapter-react-16";
import Enzyme, { shallow, mount, render } from "enzyme";
import ConnectedGame, { Game } from "../src/Game";
import BoardSize from "../src/BoardSize";
import Board from "../src/Board";
import StepList from "../src/StepList";
import configureMockStore from "redux-mock-store";
import { mapStateToProps, mapDispatchToProps } from "../src/Game";
import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";
import { reset } from "../src/actions";
import rootReducers from "../src/rootReducer";
import { createStore } from "redux";
import { act } from "react-dom/test-utils";

Enzyme.configure({ adapter: new Adapter() });

const mockStore = configureMockStore();
let rows = 15; // 8 * 8 button matrix
let win_rule = 5; // 5 adjacent buttons with same char.
const initialState = {
  reducer: {
    history: [
      {
        squares: Array(rows ** 2).fill(null),
      },
    ], // initiate state of squares of the board and keep all the steps' states of all squares.
    xIsNext: true,
    stepNumber: 0,
    rows: rows, // also is the number of columns.
    max_rows: 20, //max limit of rows.
    fixed: false, // after rows is changed in left-up input box, it is true. No more change further.
    locations: [
      {
        row: 0,
        col: 0,
      },
    ],
    winner_stepNumber: -1,
    winner: null,
    order: "Ascend",
    match: [], // hold the square indices(linear in array) when win the game.
    win_rule: win_rule,
    status: null,
    win_color: null,
    bold2: null,
  },
};
/* Not work 
const waitForComponentToPaint = async (wrapper) => {
  act(async () => {
    await new Promise((resolve) => setTimeout(resolve));
    wrapper.update();
  });
};*/

describe("Game", () => {
  let store;
  let component;
  // set up a fake store for all our tests
  beforeEach(() => {
    store = mockStore(initialState);

    store.dispatch = jest.fn();

    store.getActions = () => {
      return [{ type: "RESET" }];
    };

    component = TestRenderer.create(
      <Provider store={store}>
        <ConnectedGame rows_={15} />
      </Provider>
    );
  });

  it("should render BoardSize and Board", () => {
    const wrapper = shallow(<Game />);
    expect(
      wrapper.containsMatchingElement([<BoardSize />, <Board />, <StepList />])
    ).toEqual(true);
  });

  it("should render with given state from Redux store", () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it("should restart the game with state stepNumber == 0", () => {
    const wrapper = shallow(
      <Provider store={store}>
        <ConnectedGame rows_={15} />
      </Provider>
    );

    expect(wrapper.props().value.store.getState().reducer.stepNumber).toEqual(
      0
    );
  });

  // Simulate Cypress
  it("should reset the game with state stepNumber == 0", () => {
    let new_store = createStore(rootReducers);
    const wrapper = mount(
      <Provider store={new_store}>
        <ConnectedGame rows_={15} />
      </Provider>
    );
    expect(wrapper.find("li.step_list").length).toEqual(1);

    let squares = wrapper.find("button.square");
    let first_square = squares.get(0);
    act(() => {
      first_square.props.onClick();
    });
    wrapper.update();
    expect(wrapper.find("li.step_list").length).toEqual(2);

    let reset_wrap = wrapper.find(".reset");
    let reset_button = reset_wrap.get(0);
    act(() => {
      reset_button.props.onClick();
    });
    wrapper.update();
    expect(wrapper.find("li.step_list").length).toEqual(1);
  });

  //https://jsramblings.com/3-ways-to-test-mapstatetoprops-and-mapdispatchtoprops/
  it("test mapStateToProps", () => {
    const initialState = {
      reducer: {
        status: "Winer: O",
        bold2: "yes",
        win_color: "red",
      },
    };

    // Just call the method directly passing in sample data
    // to make sure it does what it's supposed to
    expect(mapStateToProps(initialState).status).toEqual(
      initialState.reducer.status
    );
    expect(mapStateToProps(initialState).bold2).toEqual(
      initialState.reducer.bold2
    );
    expect(mapStateToProps(initialState).win_color).toEqual(
      initialState.reducer.win_color
    );
  });

  it("test mapDispatchToPrpos", () => {
    const dispatch = jest.fn();

    // For the `mapDispatchToProps`, call it directly but pass in
    // a mock function and check the arguments passed in are as expected
    mapDispatchToProps(dispatch).reset();
    expect(dispatch.mock.calls[0][0]).toEqual({ type: "RESET" });
  });

  it("should hit restart method in Game when hit reset button", () => {
    const spy = jest.spyOn(Game.prototype, "restart");

    const wrapper = mount(
      <Provider store={store}>
        <ConnectedGame rows_={15} />
      </Provider>
    );
    // restart method should pass.
    expect(spy).not.toHaveBeenCalled();
    wrapper.find("div.game-reset button").simulate("click");
    expect(spy).toHaveBeenCalledTimes(1);
    spy.mockRestore();
  });

  it("should hit reset action when click reset button", () => {
    const reset_spy = jest.spyOn(require("../src/actions"), "reset");
    TestRenderer.act(() => {
      //component.root.findByType("button").props.onClick();
      //console.log(component.root.findByProps({ className: "reset" }).props);
      component.root.findByProps({ className: "reset" }).props.onClick();
    });

    expect(reset_spy).toHaveBeenCalledTimes(1);
    expect(reset_spy).toHaveBeenCalledWith();
    reset_spy.mockRestore();
  });

  // Simulate Cypress
  it("should win the game", () => {
    let new_store = createStore(rootReducers);
    const wrapper = mount(
      <Provider store={new_store}>
        <ConnectedGame rows_={15} />
      </Provider>
    );
    expect(wrapper.find(".winner").text()).toEqual("");

    let rs = wrapper.find("div.board-row");
    //console.log(rs.at(0).children().find("button").debug());
    let row_1_buttons = rs.at(0).children().find("button");
    let row_2_buttons = rs.at(1).children().find("button");

    act(() => {
      row_1_buttons.get(0).props.onClick();
      row_2_buttons.get(0).props.onClick();
      row_1_buttons.get(1).props.onClick();
      row_2_buttons.get(1).props.onClick();
      row_1_buttons.get(2).props.onClick();
      row_2_buttons.get(2).props.onClick();
      row_1_buttons.get(3).props.onClick();
      row_2_buttons.get(3).props.onClick();
      row_1_buttons.get(4).props.onClick();
    });
    wrapper.update();
    expect(wrapper.find(".winner").prop("style")).toEqual({
      color: "red",
      fontWeight: "bold",
    });
    expect(wrapper.find(".winner").text()).toEqual("Winner: X");
  });
});
