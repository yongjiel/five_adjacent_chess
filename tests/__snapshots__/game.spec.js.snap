// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Game should render with given state from Redux store 1`] = `
<div>
  <div
    className="game-input"
  >
    <div
      className="rows"
    >
      <input
        disabled={false}
        id="rows"
        onChange={[Function]}
        placeholder={15}
        size="10"
        style={Object {}}
        value={15}
      />
      Â 
      <b>
        rows
      </b>
       and 
      <b>
        columns
      </b>
    </div>
  </div>
   
  <br />
  <div
    className="game"
  >
    <div
      className="game-board"
    >
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
      <div
        className="board-row"
      >
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
        <button
          className="square "
          onClick={[Function]}
        >
           
        </button>
      </div>
    </div>
    <div
      className="game-info"
    >
      <div
        className="game-reset"
      >
        <ul>
          <button
            className="reset"
            onClick={[Function]}
          >
            Restart the Game
          </button>
        </ul>
      </div>
      <div
        className="winner"
        style={
          Object {
            "color": null,
            "fontWeight": null,
          }
        }
      />
      <div
        className="step-list"
      >
        <ul>
          <button
            onClick={[Function]}
          >
            <b>
              Ascend
            </b>
          </button>
        </ul>
        <ol
          reversed={false}
        >
          <li
            className="step_list"
          >
            <button
              onClick={[Function]}
              style={
                Object {
                  "color": "blue",
                  "fontWeight": "bold",
                }
              }
            >
              Ascend
              , Step: 
              0
              , current_step: 
              0
              , Row:
               
              0
               Col: 
              0
            </button>
          </li>
        </ol>
      </div>
    </div>
  </div>
</div>
`;
